### PR TITLE
[Merged by Bors] - feat: The category Fact(f)

### PR DESCRIPTION
**Pull request by javra**
This prepares to have Factorisation systems in Lean as well as to serve as a generalisation of things like `MonoFactorisation`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by javra** [comment]
Questions:
* Is `Fact f` an okay name if it's inside `CategoryTheory`, or do we still want to avoid the clash with `_root_.Fact`?
* How do we avoid the `change @Eq...`?
* Is the way I defined the initial and terminal object good style?

**Comment by jcommelin** [comment]
@javra do I understand that you want this to be reviewed? If so, I think you should switch from draft to regular PR and add `awaiting-review`. Or otherwise start a discussion on zulip. Because a draft PR without `awaiting-review` is not very visible on our Github queues, and so your questions will not be seen by many.

**Comment by javra** [comment]
@jcommelin I'm at the LftCM and wanted to collect some local feedback first, from people who started on AWFS, but I guess I could convert it to a proper PR as well :upside_down_face:

**Comment by jcommelin** [comment]
Ok, got it. Fwiw, I think the material looks good. But I do think that you'' get clashes with `_root_.Fact`. What about `FactCat`?

**Comment by javra** [comment]
I'd be happy with `FactCat`. If we want to use the type of factorisations without caring about the category structures, `Factorisation f` would probably the better type name. Just thought that since `_root_.Fact` takes propositions, there's maybe no ambiguities at all?

What about the field names of `Fact`? The `D` is just taken from nlab for the lack of a better idea...

File: Mathlib/CategoryTheory/Category/Fact.lean [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+universe v u
+
+variable {C : Type u} [Category.{v} C]
+
+/-- Factorisations of a morphism `f` as a structure, containing, one object, two morphisms,
+and the condition that their composition equals `f`. -/
+structure Fact {X Y : C} (f : X ⟶ Y) where
+  /-- The midpoint of the factorisation. -/
+  D   : C
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Category/Fact.lean:32]: Maybe, `mid` or `midPr` would work as name?

File: Mathlib/CategoryTheory/Category/Factorisation.lean [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+universe v u
+
+variable {C : Type u} [Category.{v} C]
+
+/-- Factorisations of a morphism `f` as a structure, containing, one object, two morphisms,
+and the condition that their composition equals `f`. -/
+structure Factorisation {X Y : C} (f : X ⟶ Y) where
+  /-- The midpoint of the factorisation. -/
+  mid   : C
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Category/Factorisation.lean:32]:
```suggestion
  mid  : C
```

**Review by jcommelin — APPROVED** [review]
LGTM

bors d+

**Pull request closed without merging**